### ISSUE TITLE
 Can't Write To Stream With A Previously Used Name

### ISSUE THREAD

**Issue by jdmaguire** [pravega/pravega#1681]
**Problem description**
Similar to #1012 (Can't Delete Or Seal Stream With A Previously Used Name). If I create a stream with the same name as a previously created stream (one that was deleted), I can't write to it.
 
**Reproduction**

I can consistently reproduce this when:

    I create a scope named jdm
    I create a stream in it named kyle
    I seal kyle
    I delete kyle
    Later, I create a stream called kyle

At this point, I can't write to kyle. If I create a new EventStreamWriter, I get the following exception thrown:

```
Exception in thread "main" java.lang.IllegalArgumentException: No such segment: jdm/aStream/0
	at io.pravega.client.segment.impl.SegmentOutputStreamImpl$ResponseProcessor.noSuchSegment(SegmentOutputStreamImpl.java:309)
	at io.pravega.shared.protocol.netty.WireCommands$NoSuchSegment.process(WireCommands.java:188)
	at io.pravega.client.netty.impl.ClientConnectionInboundHandler.channelRead(ClientConnectionInboundHandler.java:92)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:363)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:349)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:341)
	at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:293)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:267)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:363)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:349)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:341)
	at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:293)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:267)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:363)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:349)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:341)
	at io.netty.channel.ChannelInboundHandlerAdapter.channelRead(ChannelInboundHandlerAdapter.java:86)
	at io.pravega.shared.protocol.netty.ExceptionLoggingHandler.channelRead(ExceptionLoggingHandler.java:37)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:363)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:349)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:341)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1334)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:363)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:349)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:926)
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:129)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:642)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:565)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:479)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:441)
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:858)
	at io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:144)
	at java.lang.Thread.run(Thread.java:748)
```

**Problem location**
EventStreamWriter

1532.26ab2c5

**Suggestions for an improvement**

Let me write to the stream, don't let me create it in the first place, or if this is intentional, give me a more informative message when I try to write to it.

**Comment by fpj** [comment]
@jdmaguire has the second call to create `kyle` succeeded? If it did, the controller should have created the stream segments, so I'm not sure why it is saying that the segment doesn't exist. The possible options I see are:

- Controller is pointing to the wrong segment store server
- Segment store server uses stale information when responding to this request

It would be useful to see the logs to determine the exact sequence of operations.

**Comment by jdmaguire** [comment]
The second call to create the stream did succeed. I'll boot up Pravega in docker-compose and try to fetch the logs.

**Comment by jdmaguire** [comment]
I created a scope called jdmTesting, stream called `willDelete`. I deleted the stream. Here is the logs from all the components when creating a new stream with that name. (Ran this in docker compose.)
```
zookeeper_1    | 2018-03-22 19:34:40,165 [myid:] - INFO  [ProcessThread(sid:0 cport:2181)::PrepRequestProcessor@841] - Got user-level KeeperException when processing sessionid:0x1000016b6060000 type:create2 cxid:0x2b8 zxid:0x12d txntype:-1 reqpath:n/a Error Path:/pravega/pravega-cluster/hostTaskIndex/172.19.0.4-6d6da75a-3a34-46ff-aab3-f1ce77c12d29 Error:KeeperErrorCode = NoNode for /pravega/pravega-cluster/hostTaskIndex/172.19.0.4-6d6da75a-3a34-46ff-aab3-f1ce77c12d29
zookeeper_1    | 2018-03-22 19:34:40,200 [myid:] - INFO  [ProcessThread(sid:0 cport:2181)::PrepRequestProcessor@841] - Got user-level KeeperException when processing sessionid:0x1000016b6060000 type:create2 cxid:0x2c2 zxid:0x131 txntype:-1 reqpath:n/a Error Path:/pravega/pravega-cluster/store/jdmTesting/willDelete Error:KeeperErrorCode = NoNode for /pravega/pravega-cluster/store/jdmTesting/willDelete
zookeeper_1    | 2018-03-22 19:34:40,239 [myid:] - INFO  [ProcessThread(sid:0 cport:2181)::PrepRequestProcessor@841] - Got user-level KeeperException when processing sessionid:0x1000016b6060000 type:create2 cxid:0x2d0 zxid:0x138 txntype:-1 reqpath:n/a Error Path:/pravega/pravega-cluster/transactions/activeTx/jdmTesting/willDelete/0 Error:KeeperErrorCode = NodeExists for /pravega/pravega-cluster/transactions/activeTx/jdmTesting/willDelete/0
controller_1   | 2018-03-22 19:34:40,254 632233 [controllerpool-63] INFO  i.p.c.t.Stream.StreamMetadataTasks - jdmTesting/willDelete created in metadata store
controller_1   | 2018-03-22 19:34:40,263 632242 [epollEventLoopGroup-2-1] INFO  i.p.c.n.i.ClientConnectionInboundHandler - Connection established ChannelHandlerContext(ClientConnectionInboundHandler#0, [id: 0xdd72917e])
segmentstore_1 | 2018-03-22 19:34:40,270 590231 [epollEventLoopGroup-2-1] INFO  i.n.handler.logging.LoggingHandler - [id: 0x656ea21e, L:/172.19.0.8:12345] READ: [id: 0xa91c031f, L:/172.19.0.8:12345 - R:/172.19.0.1:47368]
segmentstore_1 | 2018-03-22 19:34:40,271 590232 [epollEventLoopGroup-2-1] INFO  i.n.handler.logging.LoggingHandler - [id: 0x656ea21e, L:/172.19.0.8:12345] READ COMPLETE
controller_1   | 2018-03-22 19:34:40,268 632247 [epollEventLoopGroup-2-2] INFO  i.p.c.n.i.ClientConnectionInboundHandler - Connection established ChannelHandlerContext(ClientConnectionInboundHandler#0, [id: 0x0b2a39c4])
segmentstore_1 | 2018-03-22 19:34:40,277 590238 [epollEventLoopGroup-2-1] INFO  i.n.handler.logging.LoggingHandler - [id: 0x656ea21e, L:/172.19.0.8:12345] READ: [id: 0xca43b967, L:/172.19.0.8:12345 - R:/172.19.0.1:47370]
segmentstore_1 | 2018-03-22 19:34:40,277 590238 [epollEventLoopGroup-2-1] INFO  i.n.handler.logging.LoggingHandler - [id: 0x656ea21e, L:/172.19.0.8:12345] READ COMPLETE
segmentstore_1 | 2018-03-22 19:34:40,283 590244 [epollEventLoopGroup-3-1] INFO  i.p.s.s.h.handler.AppendProcessor - Received hello from connection: [id: 0xca43b967, L:/172.19.0.8:12345 - R:/172.19.0.1:47370]
segmentstore_1 | 2018-03-22 19:34:40,284 590245 [epollEventLoopGroup-3-1] INFO  i.p.s.s.h.h.PravegaRequestProcessor - Segment 'jdmTesting/willDelete/2' already exists and cannot perform operation 'Create segment'.
controller_1   | 2018-03-22 19:34:40,287 632266 [epollEventLoopGroup-2-1] INFO  i.p.c.n.i.ClientConnectionInboundHandler - Connection established ChannelHandlerContext(ClientConnectionInboundHandler#0, [id: 0xdca0832a])
segmentstore_1 | 2018-03-22 19:34:40,290 590251 [epollEventLoopGroup-2-1] INFO  i.n.handler.logging.LoggingHandler - [id: 0x656ea21e, L:/172.19.0.8:12345] READ: [id: 0xf2f6becc, L:/172.19.0.8:12345 - R:/172.19.0.1:47372]
segmentstore_1 | 2018-03-22 19:34:40,292 590253 [epollEventLoopGroup-2-1] INFO  i.n.handler.logging.LoggingHandler - [id: 0x656ea21e, L:/172.19.0.8:12345] READ COMPLETE
controller_1   | 2018-03-22 19:34:40,302 632281 [epollEventLoopGroup-2-2] INFO  i.p.s.p.netty.FailingReplyProcessor - Received hello: WireCommands.Hello(type=HELLO, highVersion=4, lowVersion=1)
controller_1   | 2018-03-22 19:34:40,306 632285 [epollEventLoopGroup-2-2] INFO  i.p.controller.server.SegmentHelper - CreateSegment jdmTesting/willDelete/2 segmentAlreadyExists
segmentstore_1 | 2018-03-22 19:34:40,308 590269 [epollEventLoopGroup-3-2] INFO  i.p.s.s.h.handler.AppendProcessor - Received hello from connection: [id: 0xa91c031f, L:/172.19.0.8:12345 - R:/172.19.0.1:47368]
segmentstore_1 | 2018-03-22 19:34:40,310 590271 [epollEventLoopGroup-3-2] INFO  i.p.s.s.h.h.PravegaRequestProcessor - Segment 'jdmTesting/willDelete/0' already exists and cannot perform operation 'Create segment'.
segmentstore_1 | 2018-03-22 19:34:40,313 590274 [epollEventLoopGroup-3-2] INFO  i.p.s.s.h.handler.AppendProcessor - Received hello from connection: [id: 0xf2f6becc, L:/172.19.0.8:12345 - R:/172.19.0.1:47372]
segmentstore_1 | 2018-03-22 19:34:40,316 590277 [epollEventLoopGroup-3-2] INFO  i.p.s.s.h.h.PravegaRequestProcessor - Segment 'jdmTesting/willDelete/1' already exists and cannot perform operation 'Create segment'.
controller_1   | 2018-03-22 19:34:40,318 632297 [epollEventLoopGroup-2-2] WARN  i.p.controller.server.SegmentHelper - CreateSegment jdmTesting/willDelete/2 Connection dropped
controller_1   | 2018-03-22 19:34:40,320 632299 [epollEventLoopGroup-2-1] INFO  i.p.s.p.netty.FailingReplyProcessor - Received hello: WireCommands.Hello(type=HELLO, highVersion=4, lowVersion=1)
controller_1   | 2018-03-22 19:34:40,322 632301 [epollEventLoopGroup-2-1] INFO  i.p.controller.server.SegmentHelper - CreateSegment jdmTesting/willDelete/0 segmentAlreadyExists
controller_1   | 2018-03-22 19:34:40,323 632302 [epollEventLoopGroup-2-1] INFO  i.p.s.p.netty.FailingReplyProcessor - Received hello: WireCommands.Hello(type=HELLO, highVersion=4, lowVersion=1)
controller_1   | 2018-03-22 19:34:40,324 632303 [epollEventLoopGroup-2-1] INFO  i.p.controller.server.SegmentHelper - CreateSegment jdmTesting/willDelete/1 segmentAlreadyExists
controller_1   | 2018-03-22 19:34:40,330 632309 [epollEventLoopGroup-2-1] WARN  i.p.controller.server.SegmentHelper - CreateSegment jdmTesting/willDelete/0 Connection dropped
controller_1   | 2018-03-22 19:34:40,331 632310 [epollEventLoopGroup-2-1] WARN  i.p.controller.server.SegmentHelper - CreateSegment jdmTesting/willDelete/1 Connection dropped
controller_1   | 2018-03-22 19:34:40,360 632339 [controllerpool-35] INFO  i.p.c.s.r.r.StreamMetadataResourceImpl - Successfully created stream: jdmTesting/willDelete
segmentstore_1 | 2018-03-22 19:34:43,593 593554 [core-27] INFO  i.p.s.s.h.ZKSegmentContainerMonitor - Container Changes: Desired = [0, 1, 2, 3], Current = [0, 1, 2, 3], PendingTasks = [], ToStart = [], ToStop = [].
segmentstore_1 | 2018-03-22 19:34:53,595 603556 [core-28] INFO  i.p.s.s.h.ZKSegmentContainerMonitor - Container Changes: Desired = [0, 1, 2, 3], Current = [0, 1, 2, 3], PendingTasks = [], ToStart = [], ToStop = [].
 
```

**Comment by jdmaguire** [comment]
Commands I ran to get this:

```
$ set HOST_IP <ip> docker-compose up
```

and:

```
$ set host localhost:10080 #export host=localhost:10080 in bash
$ curl -X POST -H "content-type: application/json" -d '{"scopeName": "jdmTesting"}' $host/v1/scopes
$ curl -X POST -H "content-type: application/json" -d '{"streamName":"willDelete","scopeName":"jdmtesting","scalingPolicy":{"type":"BY_RATE_IN_KBYTES_PER_SEC","targetRate":5120,"scaleFactor":1,"minSegments":3}}' $host/v1/scopes/jdmtesting/streams -v
$ curl -X DELETE -H "content-type: application/json" -d '{"streamState":"SEALED"}' $host/v1/scopes/jdmTesting/streams/willDelete -v
$ sleep 120
$ curl -X POST -H "content-type: application/json" -d '{"streamName":"willDelete","scopeName":"jdmtesting","scalingPolicy":{"type":"BY_RATE_IN_KBYTES_PER_SEC","targetRate":5120,"scaleFactor":1,"minSegments":3}}' $host/v1/scopes/jdmTesting/streams -v
$ curl -X DELETE $host/v1/scopes/jdmTesting/streams/willDelete -v #fails
$ curl -X PUT -H "content-type: application/json" -d '{"streamState":"SEALED"}' $host/v1/scopes/jdmTesting/streams/willDelete/state -v #fails
```

**Comment by RaulGracia** [comment]
Closed with #2800.
